### DISCUSSION TITLE
Marbles

### DISCUSSION THREAD

**Discussion by vxgmichel** [tomhea/flip-jump#214] [Show and tell]
Hi :wave:

I made an esoteric programming language based on marble circuitry called [marbles](https://github.com/vxgmichel/marbles) and FlipJump has been a huge help in demonstrating that the language is Turing-complete so thank you for that :)

I wrote a bunch of information about that [over there](https://github.com/vxgmichel/marbles#flipjump-computers).

 There are two things that are specific to the FlipJump computers I'm generating with the [flipjump-to-marbles.py](https://github.com/vxgmichel/marbles/blob/main/flipjump-to-marbles.py) script:
- The jump addresses are always expected to be instruction aligned, meaning the least significant bits in each jump address are not even represented.
- The bit that are meant to be toggled for writing `0` and `1` to the output bit stream (bit 0 and 1 in the 3rd word) always return `0` when read.

Would those points break existing flip jump programs or are they reasonable to take for granted?

**Comment by tomhea** [comment]
Hi, first of all nice idea!

About your points:
* Yes, the default flipjump cpu expects the jump address to fall exactly on a new word (dw-aligned).
* That is not an assumption that I've made. Right now reading those output bits - you'll read their value as they were flipped, like any other bit. Is it something that you can handle with?

I'm happy that you find flipjump usefull in your project :)

**Comment by vxgmichel** [comment]
> Yes, the default flipjump cpu expects the jump address to fall exactly on a new word (dw-aligned).

Oh right I didn't check the cpp implementation, I just had a look at the python one which seems to allow it.

> That is not an assumption that I've made. Right now reading those output bits - you'll read their value as they were flipped, like any other bit. Is it something that you can handle with?

It is definitely doable, it's just that it's more logic to pack in an already tight space. Since jumping to this address is usually done to check the input bit, I thought I could simply ignore the toggle operation and get away with it ^^

So far the flip jump programs I tried still work as expected but I'll try to fix that in the future.

Thanks for your feedback :)

**Reply by tomhea** [reply]
Oh I get you.
I assumed that the IO+0/1 flips will only affect the [jump to input], so that it will flip bits 0/1/2/3 instead of flipping 0 before its jump. So if you dont go back to address 0 - it shouldn't be a problem.

**Reply by tomhea** [reply]
Btw, if enough .fj programs will be available on github - it could get recognized as a language by github, so feal free to write as many programs as you need :)
